File: exam_LRX/py/work10.py
# 10.给出图像处理中频域滤波法中理想低通滤波器、Butterworth低通滤波器、高斯低通滤波器实例各一个。
import cv2
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams

# 设置中文字体
rcParams['font.family'] = 'SimHei'  # 设置为黑体
rcParams['axes.unicode_minus'] = False  # 解决负号显示问题

# 定义频域滤波器函数
def ideal_lowpass_filter(shape, cutoff):
    """理想低通滤波器"""
    rows, cols = shape
    center = (rows // 2, cols // 2)
    filter_mask = np.zeros((rows, cols), dtype=np.float32)
    for i in range(rows):
        for j in range(cols):
            distance = np.sqrt((i - center[0]) ** 2 + (j - center[1]) ** 2)
            if distance <= cutoff:
                filter_mask[i, j] = 1
    return filter_mask

def butterworth_lowpass_filter(shape, cutoff, order=2):
    """Butterworth低通滤波器"""
    rows, cols = shape
    center = (rows // 2, cols // 2)
    filter_mask = np.zeros((rows, cols), dtype=np.float32)
    for i in range(rows):
        for j in range(cols):
            distance = np.sqrt((i - center[0]) ** 2 + (j - center[1]) ** 2)
            filter_mask[i, j] = 1 / (1 + (distance / cutoff) ** (2 * order))
    return filter_mask

def gaussian_lowpass_filter(shape, cutoff):
    """高斯低通滤波器"""
    rows, cols = shape
    center = (rows // 2, cols // 2)
    filter_mask = np.zeros((rows, cols), dtype=np.float32)
    for i in range(rows):
        for j in range(cols):
            distance = np.sqrt((i - center[0]) ** 2 + (j - center[1]) ** 2)
            filter_mask[i, j] = np.exp(-(distance ** 2) / (2 * (cutoff ** 2)))
    return filter_mask

# 应用频域滤波
def apply_filter(image, filter_mask):
    """对图像应用频域滤波"""
    dft = np.fft.fft2(image)
    dft_shifted = np.fft.fftshift(dft)
    filtered_dft = dft_shifted * filter_mask
    dft_inverse = np.fft.ifftshift(filtered_dft)
    filtered_image = np.abs(np.fft.ifft2(dft_inverse))
    return filtered_image

# 读取灰度图像
image = cv2.imread('../image/image9.png', cv2.IMREAD_GRAYSCALE)

# 定义滤波器参数
cutoff = 50
butterworth_order = 2

# 生成滤波器并应用
ideal_filter = ideal_lowpass_filter(image.shape, cutoff)
butterworth_filter = butterworth_lowpass_filter(image.shape, cutoff, butterworth_order)
gaussian_filter = gaussian_lowpass_filter(image.shape, cutoff)

ideal_result = apply_filter(image, ideal_filter)
butterworth_result = apply_filter(image, butterworth_filter)
gaussian_result = apply_filter(image, gaussian_filter)

# 显示结果
plt.figure(figsize=(12, 8))
plt.subplot(1, 4, 1), plt.imshow(image, cmap='gray')
plt.title("原图"), plt.axis('off')
plt.subplot(1, 4, 2), plt.imshow(ideal_result, cmap='gray')
plt.title("理想滤波图"), plt.axis('off')
plt.subplot(1, 4, 3), plt.imshow(butterworth_result, cmap='gray')
plt.title("Butterworth滤波图"), plt.axis('off')
plt.subplot(1, 4, 4), plt.imshow(gaussian_result, cmap='gray')
plt.title("高斯滤波图"), plt.axis('off')

plt.tight_layout()
plt.show()

File: exam_LRX/cpp/work10.cpp
//
// Created by liryi on 24-11-22.
// 10.给出图像处理中频域滤波法中理想低通滤波器、Butterworth低通滤波器、高斯低通滤波器实例各一个。
#include <opencv2/opencv.hpp>
#include <iostream>
#include <cmath>

using namespace cv;
using namespace std;

// 频谱中心转换
void shiftDFT(Mat& src, Mat& dst) {
    dst = src.clone();
    int cx = dst.cols / 2;
    int cy = dst.rows / 2;

    Mat q0(dst, Rect(0, 0, cx, cy));
    Mat q1(dst, Rect(cx, 0, cx, cy));
    Mat q2(dst, Rect(0, cy, cx, cy));
    Mat q3(dst, Rect(cx, cy, cx, cy));

    Mat tmp;
    q0.copyTo(tmp);
    q3.copyTo(q0);
    tmp.copyTo(q3);

    q1.copyTo(tmp);
    q2.copyTo(q1);
    tmp.copyTo(q2);
}

// 理想低通滤波器
Mat idealLowpassFilter(Size size, int cutoff) {
    Mat filter(size, CV_32F, Scalar(0));
    Point center(size.width / 2, size.height / 2);

    for (int i = 0; i < size.height; ++i) {
        for (int j = 0; j < size.width; ++j) {
            double distance = sqrt(pow(i - center.y, 2) + pow(j - center.x, 2));
            if (distance <= cutoff) {
                filter.at<float>(i, j) = 1.0;
            }
        }
    }
    return filter;
}

// Butterworth低通滤波器
Mat butterworthLowpassFilter(Size size, int cutoff, int order) {
    Mat filter(size, CV_32F, Scalar(0));
    Point center(size.width / 2, size.height / 2);

    for (int i = 0; i < size.height; ++i) {
        for (int j = 0; j < size.width; ++j) {
            double distance = sqrt(pow(i - center.y, 2) + pow(j - center.x, 2));
            filter.at<float>(i, j) = 1.0 / (1.0 + pow(distance / cutoff, 2 * order));
        }
    }
    return filter;
}

// 高斯低通滤波器
Mat gaussianLowpassFilter(Size size, int cutoff) {
    Mat filter(size, CV_32F, Scalar(0));
    Point center(size.width / 2, size.height / 2);

    for (int i = 0; i < size.height; ++i) {
        for (int j = 0; j < size.width; ++j) {
            double distance = sqrt(pow(i - center.y, 2) + pow(j - center.x, 2));
            filter.at<float>(i, j) = exp(-(pow(distance, 2)) / (2 * pow(cutoff, 2)));
        }
    }
    return filter;
}

// 应用滤波器
Mat applyFilter(const Mat& image, const Mat& filter, Mat& spectrumBefore, Mat& spectrumAfter) {
    Mat padded;
    int m = getOptimalDFTSize(image.rows);
    int n = getOptimalDFTSize(image.cols);
    copyMakeBorder(image, padded, 0, m - image.rows, 0, n - image.cols, BORDER_CONSTANT, Scalar::all(0));

    Mat planes[] = { Mat_<float>(padded), Mat::zeros(padded.size(), CV_32F) };
    Mat complexImage;
    merge(planes, 2, complexImage);

    // 傅里叶变换
    dft(complexImage, complexImage);

    // 转换到频谱中心
    Mat shifted;
    shiftDFT(complexImage, shifted);

    // 保存原始频谱
    Mat planesSplit[2];
    split(shifted, planesSplit); // 分离实部和虚部
    magnitude(planesSplit[0], planesSplit[1], spectrumBefore); // 计算幅值
    spectrumBefore += Scalar::all(1); // 避免 log(0)
    log(spectrumBefore, spectrumBefore); // 转换到对数尺度

    // 滤波器应用
    planesSplit[0] = planesSplit[0].mul(filter);
    planesSplit[1] = planesSplit[1].mul(filter);
    merge(planesSplit, 2, shifted);

    // 保存滤波后的频谱
    split(shifted, planesSplit);
    magnitude(planesSplit[0], planesSplit[1], spectrumAfter);
    spectrumAfter += Scalar::all(1);
    log(spectrumAfter, spectrumAfter);

    // 逆傅里叶变换
    shiftDFT(shifted, complexImage);
    idft(complexImage, complexImage, DFT_SCALE | DFT_REAL_OUTPUT);
    Mat result;
    complexImage(Rect(0, 0, image.cols, image.rows)).convertTo(result, CV_8U);

    return result;
}

// 在一个窗口中显示原图和滤波后的图像
void displayFilters(const Mat& originalImage, const Mat& idealImage, const Mat& butterworthImage, const Mat& gaussianImage) {
    int width = originalImage.cols;
    int height = originalImage.rows;
    Mat display = Mat::zeros(height, width * 4, originalImage.type());

    originalImage.copyTo(display(Rect(0, 0, width, height)));
    putText(display, "Original", Point(10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    idealImage.copyTo(display(Rect(width, 0, width, height)));
    putText(display, "Ideal", Point(width + 10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    butterworthImage.copyTo(display(Rect(width * 2, 0, width, height)));
    putText(display, "Butterworth", Point(width * 2 + 10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    gaussianImage.copyTo(display(Rect(width * 3, 0, width, height)));
    putText(display, "Gaussian", Point(width * 3 + 10, 30), FONT_HERSHEY_SIMPLEX, 1, Scalar(255), 2);

    namedWindow("Filters", WINDOW_NORMAL);
    imshow("Filters", display);
    waitKey(0);
}

int main() {
    Mat image = imread("E:\\DL_class\\exam_LRX\\cpp\\images\\image9.png", IMREAD_GRAYSCALE);
    if (image.empty()) {
        cout << "Error: Image not found!" << endl;
        return -1;
    }

    int cutoff = 50;
    int order = 2;

    Mat idealFilter = idealLowpassFilter(image.size(), cutoff);
    Mat butterworthFilter = butterworthLowpassFilter(image.size(), cutoff, order);
    Mat gaussianFilter = gaussianLowpassFilter(image.size(), cutoff);

    Mat idealResult, butterworthResult, gaussianResult;
    Mat idealSpectrumBefore, idealSpectrumAfter;
    Mat butterworthSpectrumBefore, butterworthSpectrumAfter;
    Mat gaussianSpectrumBefore, gaussianSpectrumAfter;

    idealResult = applyFilter(image, idealFilter, idealSpectrumBefore, idealSpectrumAfter);
    butterworthResult = applyFilter(image, butterworthFilter, butterworthSpectrumBefore, butterworthSpectrumAfter);
    gaussianResult = applyFilter(image, gaussianFilter, gaussianSpectrumBefore, gaussianSpectrumAfter);

    // 显示结果
    displayFilters(image, idealResult, butterworthResult, gaussianResult);

    return 0;
}



